### ISSUE TITLE
Test failures when number of CPU is 1

### ISSUE THREAD

**Issue by sanvila** [joblib/joblib#1618]
Hello. As reported in Debian:

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1085692

Some tests fail when the machine has a single CPU:

```
=========================== short test summary info ============================
FAILED joblib/test/test_parallel.py::test_nested_parallelism_limit[parallel_config-loky] - AssertionError: assert [('LokyBacken...lBackend', 2)] == [('LokyBacken...lB...
FAILED joblib/test/test_parallel.py::test_nested_parallelism_limit[parallel_config-threading] - AssertionError: assert [('ThreadingB...lBackend', 2)] == [('ThreadingB...lB...
FAILED joblib/test/test_parallel.py::test_nested_parallelism_limit[parallel_backend-loky] - AssertionError: assert [('LokyBacken...lBackend', 2)] == [('LokyBacken...lB...
FAILED joblib/test/test_parallel.py::test_nested_parallelism_limit[parallel_backend-threading] - AssertionError: assert [('ThreadingB...lBackend', 2)] == [('ThreadingB...lB...
FAILED joblib/test/test_parallel.py::test_threadpool_limitation_in_child_override[parallel_config-OPENBLAS_NUM_THREADS--1] - AssertionError: assert ['4', '4'] == ['1', '1']
FAILED joblib/test/test_parallel.py::test_threadpool_limitation_in_child_override[parallel_config-MKL_NUM_THREADS--1] - AssertionError: assert ['4', '4'] == ['1', '1']
FAILED joblib/test/test_parallel.py::test_threadpool_limitation_in_child_override[parallel_config-OMP_NUM_THREADS--1] - AssertionError: assert ['4', '4'] == ['1', '1']
FAILED joblib/test/test_parallel.py::test_threadpool_limitation_in_child_override[parallel_backend-OPENBLAS_NUM_THREADS--1] - AssertionError: assert ['4', '4'] == ['1', '1']
FAILED joblib/test/test_parallel.py::test_threadpool_limitation_in_child_override[parallel_backend-MKL_NUM_THREADS--1] - AssertionError: assert ['4', '4'] == ['1', '1']
FAILED joblib/test/test_parallel.py::test_threadpool_limitation_in_child_override[parallel_backend-OMP_NUM_THREADS--1] - AssertionError: assert ['4', '4'] == ['1', '1']
= 10 failed, 1308 passed, 42 skipped, 2 deselected, 2 xfailed, 4 xpassed, 317 warnings in 84.09s (0:01:24) =
```

This used to work in version 0.13.0, and there are occurrences of `cpu_count() == 1` in the code, so I believe this was not supposed to happen.

Thanks.

**Comment by lesteve** [comment]
I agree the tests with `cpu_count` looks like they should handle this [^1] so I would guess that `cpu_count()` is wrong in this case.

Is it doable to double-check what `cpu_count()` returns in the Debian CI?

Do you happen to know how the restriction to 1 CPU is done, for example `cgroup`, CPU affinity, something else? This would help trying to reproduce and maybe fix `cpu_count`.

[^1]: other tests would need to be adapted I think

**Comment by sanvila** [comment]
The error does not happen on Debian CI, it happens when trying to build the package on a virtual machine with a single-CPU (for example, AWS instances of type m7a.medium or  r7a.medium). Building the archive on single-CPU machines is still the most efficient way to build all Debian packages.

If you don't have a single-CPU virtual machine to test, `GRUB_CMDLINE_LINUX="nr_cpus=1"` usually yields the same outcome.

In the meantime, we fixed the Debian package in this commit:

https://salsa.debian.org/science-team/joblib/-/commit/89e386f951c43a7628d6cb19675540f4c588b699

but I agree that the portable way would have been to use a skipif using `os.cpu_count()`.

Thanks.

**Comment by lesteve** [comment]
Thanks for your answer. So I think the issue is that `cpu_count()` is wrong in some specific cases to be debugged. I would have thought that, with our cgroup handling, we were covering AWS instances but maybe not ...

I'll try to give a go to `GRUB_CMDLINE="nr_cpus=1"`.

**Comment by lesteve** [comment]
Turns out, it is an issue with the tests themselves when `cpu_count() == 1` rather than `cpu_count` being wrong.

I can reproduce the failures by running the following command:
```
LOKY_MAX_CPU_COUNT=1 pytest joblib/test/test_parallel.py
```

**Comment by lesteve** [comment]
After having looked a bit more, this is not a problem in `joblib` but more than the tests needs to be updated for the `cpu_count() == 1` use case.

I guess it may also make sense to have a CI build with `LOKY_MAX_CPU_COUNT=1` so that we avoid creating issues in Debian packaging ...

**Comment by sanvila** [comment]
> I guess it may also may make sense to have a CI build with LOKY_MAX_CPU_COUNT=1 so that we avoid creating issues in Debian packaging ...

Well, as I said above, this is not a CI build in the usual sense that we can tweak adding specific extra variables. This is me trying to see what would happen if the end user tries to build the package from source following established Debian standards. Because our definition of build-essential does not mandate that the machine has to be multi-core, the package should build regardless of the number of CPUs.

The idea is that anybody can rebuild the package without doing anything special.

As for creating issues in Debian packaging, the problem is solved on the Debian side. because we added an ifeq in debian/rules. But it would be selfish on our part to just disable the test and be quiet about it. We prefer to tell upstream so that we can find a solution which is valid for everybody, not just Debian users.
